src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -74,36 +86,53 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
   // client should use "" instead.
   assert(msg != nullptr, "enqueuing a null message!");
 
-  if (!_buffer->push_back(output, decorations, msg)) {
+  size_t msg_len = strlen(msg);
+
+  if (_buffer->push_back(output, decorations, msg, msg_len)) {
+    _data_available = true;
+    _inner_lock.notify();
+    return;
+  }
+
+  if (LogConfiguration::async_mode() == LogConfiguration::AsyncMode::Stall) {
+    void* ptr = os::malloc(Message::calc_size(msg_len), mtLogging);
+    new (ptr) Message(output, decorations, msg, msg_len);
+    _stalled_message = (Message*)ptr;
+    while (_stalled_message != nullptr) {
+      _inner_lock.wait(0 /* no timeout */);
+    }
+    os::free((Message*)ptr);
+  } else {
     bool p_created;
     uint32_t* counter = _stats.put_if_absent(output, 0, &p_created);
     *counter = *counter + 1;
-    return;
   }
-
-  _data_available = true;
-  _lock.notify();
 }
 
 void AsyncLogWriter::enqueue(LogFileStreamOutput& output, const LogDecorations& decorations, const char* msg) {
-  AsyncLogLocker locker;
+  OuterLocker locker;
+  InnerLocker ilocker;
   enqueue_locked(&output, decorations, msg);
```

Comment:
The need for the consumer lock is part of the internal implementation so I would expect to only see it acquired inside `enqueue_locked`. 

Your `InnerLocker` wrapper should expose the monitor methods as well so that all interaction is via the wrapper.